### PR TITLE
fix: totilematrixlimits

### PR DESCRIPTION
**Pull request by vitaligi**
- fix the implementation of `toTileMatrixLimits` generator
- test geometries constructors
- test `toTileMatrixLimits`

### CODE REVIEW

**Comment by vitaligi** [comment]
`line` & `polygon` tests are wip

**Review by shimoncohen — CHANGES_REQUESTED** [review]

File: src/geometries/geometry.ts [hunk 1]
```diff
@@ -57,12 +55,12 @@ export abstract class Geometry<G extends GeoJSONGeometry> {
...
   /**
-   * Find the minimal bounding tile containing the bounding box
+   * Find the tile index of minimal bounding tile containing the bounding box
    * @param tileMatrixSet tile matrix set for the containing tile lookup
    * @param metatile size of a metatile
-   * @returns tile that fully contains the bounding box in a single tile or null if it could not be fully contained in any tile
+   * @returns tile index of a tile that fully contains the bounding box in a single tile or null if it could not be fully contained in any tile
    */
-  public minimalBoundingTile<T extends TileMatrixSet>(tileMatrixSet: T, metatile = 1): Tile<T> | null {
+  public findMinimalBoundingTileIndex<T extends TileMatrixSet>(tileMatrixSet: T, metatile = 1): TileIndex<T> | null {
```
[the remark this answers is not in the document]
Comment (vitaligi) [reply src/geometries/geometry.ts:63]: i thought of mimicking the behavior of Array.prototype.find() (although it returns `undefined` perhaps i should change the return type). the addition of `find...` to func name is an attempt to convey that a result could also be not found (`null`)

File: src/tiles/tileMatrixSet.ts [hunk 2]
```diff
@@ -16,7 +16,7 @@ export class TileMatrixSet implements TileMatrixSetType {
   private readonly tileMatrixSet: TileMatrixSetType;
   public constructor(tileMatrixSetJSON: TileMatrixSetJSON) {
     // validateTileMatrixSet(tileMatrixSet); // TODO: missing implementation
-    validateCRS(tileMatrixSetJSON.crs);
+    // validateCRS(tileMatrixSetJSON.crs); // TODO: needs a different validation
```
[the remark this answers is not in the document]
Comment (vitaligi) [reply src/tiles/tileMatrixSet.ts:19]: it is indeed not needed to be here and should be tested inside validateTileMatrixSet (or should be named validateTileMatrixSetJSON)

File: src/validations/validations.ts [hunk 3]
```diff
@@ -89,20 +123,21 @@ export function validateMetatile(metatile: number): void {
  * @param tileMatrix the tile matrix to validate
  */
 export function validateTileMatrix(tileMatrix: TileMatrix): void {
-  if (tileMatrix.matrixWidth < 1) {
-    throw new Error('width of tile matrix must be at least 1');
+  const { matrixHeight, matrixWidth, tileHeight, tileWidth } = tileMatrix;
+  if (matrixWidth < 1 && Number.isSafeInteger(matrixWidth)) {
```
[the remark this answers is not in the document]
Comment (vitaligi) [reply src/validations/validations.ts:127]: done

File: src/validations/validations.ts [hunk 4]
```diff
@@ -74,13 +89,32 @@ export function validatePointByTileMatrix(point: Point, tileMatrix: TileMatrix):
...
+ * @param position position to validate
+ * @param tileMatrix tile matrix to validate `position` against
+ */
+export function validatePositionByTileMatrix(position: Position, tileMatrix: TileMatrix): void {
+  const [east, north] = position;
+  const [tileMatrixBoundingBoxMinEast, tileMatrixBoundingBoxMinNorth, tileMatrixBoundingBoxMaxEast, tileMatrixBoundingBoxMaxNorth] =
+    tileMatrixToBBox(tileMatrix);
+
+  if (east < tileMatrixBoundingBoxMinEast || east > tileMatrixBoundingBoxMaxEast) {
+    throw new RangeError(`point's easting, ${east}, is out of range of tile matrix bounding box of tile matrix: ${tileMatrix.identifier.code}`);
```
[the remark this answers is not in the document]
Comment (vitaligi) [reply src/validations/validations.ts:103]: ok replacing with something similar

File: src/validations/validations.ts [hunk 5]
```diff
@@ -74,13 +89,32 @@ export function validatePointByTileMatrix(point: Point, tileMatrix: TileMatrix):
...
+  }
+
+  if (north < tileMatrixBoundingBoxMinNorth || north > tileMatrixBoundingBoxMaxNorth) {
+    throw new RangeError(`point's northing, ${north}, is out of range of tile matrix bounding box of tile matrix: ${tileMatrix.identifier.code}`);
```
[the remark this answers is not in the document]
Comment (vitaligi) [reply src/validations/validations.ts:107]: same

File: src/validations/validations.ts [hunk 6]
```diff
@@ -1,42 +1,57 @@
...
+ */
+export function validateBBox(bbox: BBox): void {
+  const [, minNorth, , maxNorth] = bbox;
 
-  try {
-    validatePointByTileMatrix(minPoint, tileMatrix);
-    validatePointByTileMatrix(maxPoint, tileMatrix);
-  } catch (err) {
-    throw new RangeError(`bounding box is not within the tile matrix`);
+  if (maxNorth < minNorth) {
```
[the remark this answers is not in the document]
Comment (vitaligi) [reply src/validations/validations.ts:19]: you mean validation of antimeridian crossing? if so, then no
but in the context of bounding box it is not an issue as bounding box is can describe a (multi)geometry crossing the antimeridian

File: tests/geometries/helpers/tiles.ts [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+      matrixHeight,
+      matrixWidth,
+      pointOfOrigin: [eastOrigin, northOrigin],
+    } = tileMatrixJSON;
+
+    const tileMatrixHeight = cellSize * tileHeight * matrixHeight;
+    const tileMatrixWidth = cellSize * tileWidth * matrixWidth;
+
+    const [minNorth, maxNorth] =
+      cornerOfOrigin === 'topLeft' ? [northOrigin - tileMatrixHeight, northOrigin] : [northOrigin, northOrigin + tileMatrixHeight];
```
[the remark this answers is not in the document]
Comment (vitaligi) [reply tests/geometries/helpers/tiles.ts]: `cornerOfOrigin` is connected to `pointOfOrigin`. it describes the southernmost point (bottomLeft) or northernmost (topLeft)

File: tests/geometries/helpers/geometries.ts [hunk 8]
```diff
@@ -0,0 +1,37 @@
+import { fc } from '@fast-check/jest';
+import type { BBox, Position } from 'geojson';
+import { SUPPORTED_CRS } from '../../../src/constants';
+import type { GeoJSONBaseGeometry, LineStringInput, PointInput } from '../../../src/geometries/types';
```
[the remark this answers is not in the document]
Comment (vitaligi) [reply tests/geometries/helpers/geometries.ts]: outdated

File: tests/geometries/helpers/geometries.ts [hunk 9]
```diff
@@ -0,0 +1,37 @@
+import { fc } from '@fast-check/jest';
+import type { BBox, Position } from 'geojson';
```
[the remark this answers is not in the document]
Comment (vitaligi) [reply tests/geometries/helpers/geometries.ts:2]: outdated

File: tests/geometries/helpers/tiles.ts [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+    } = tileMatrix;
+    return comparedTileMatrixId === tileMatrixId;
+  });
+};
+
+export const tileMatrixToBBox = (tileMatrixSetJSON: fc.Arbitrary<TileMatrixSetJSON>, tileMatrixId: fc.Arbitrary<string>): fc.Arbitrary<BBox> => {
+  return fc.tuple(tileMatrixSetJSON, tileMatrixId).map(([tileMatrixSetJSON, tileMatrixId]) => {
+    const tileMatrixJSON = tileMatrixSetJSON.tileMatrices.find(({ id: comparedTileMatrixId }) => {
+      return comparedTileMatrixId === tileMatrixId;
+    });
```
[the remark this answers is not in the document]
Comment (vitaligi) [reply tests/geometries/helpers/tiles.ts]: i'll move the function to use `tileMatrixToBBox` implementation under `./src/tiles/tiles.ts`

File: src/validations/validations.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (shimoncohen) [reply src/validations/validations.ts:19]: Ok, can you please open an issue describing this issue in detail? Including the need to split `geometries`.

File: tests/geometries/helpers/tiles.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (shimoncohen) [reply tests/geometries/helpers/tiles.ts]: Sounds good.

File: src/geometries/geometry.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (shimoncohen) [reply src/geometries/geometry.ts:63]: I understand, but the declared return type is what should indicate that.

File: tests/geometries/helpers/geometries.ts [hunk 11]
```diff
@@ -35,3 +55,80 @@ export const generatePointInput = ({ bBox, coordinates, crs }: GenerateGeometryI
...
+      bbox: fc.option(fc.constant(coordinatesToBBox(coordinates)), { nil: undefined }),
+      coordRefSys: fc.option(crs ?? fc.constantFrom(...SUPPORTED_CRS), { nil: undefined }),
+    })
+  );
+
+  return lineInput;
+};
+
+// Notice: currently does not generates holes
+export const generatePolygonInput = ({
```
Comment (shimoncohen) [inline tests/geometries/helpers/geometries.ts:95]: Please add a constraint so we could also generate actually valid polygons (first and last point are equal).

**Review by shimoncohen — CHANGES_REQUESTED** [review]

File: src/validations/validations.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (vitaligi) [reply src/validations/validations.ts:19]: [validations issue](https://github.com/MapColonies/flat-earth/issues/24)
updating this one

File: tests/geometries/helpers/geometries.ts (hunk 11 above)
Comment (vitaligi) [reply tests/geometries/helpers/geometries.ts:99]: done

File: src/tiles/tileMatrixSet.ts [hunk 12]
```diff
@@ -16,7 +16,6 @@ export class TileMatrixSet implements TileMatrixSetType {
   private readonly tileMatrixSet: TileMatrixSetType;
   public constructor(tileMatrixSetJSON: TileMatrixSetJSON) {
     // validateTileMatrixSet(tileMatrixSet); // TODO: missing implementation
-    // validateCRS(tileMatrixSetJSON.crs); // TODO: needs a different validation
```
[the remark this answers is not in the document]
Comment (vitaligi) [reply src/tiles/tileMatrixSet.ts]: we do, but it should be done as part of `tileMatrixSetJSON` validation

File: src/validations/validations.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (vitaligi) [reply src/validations/validations.ts:19]: check

File: tests/geometries/helpers/geometries.ts [hunk 13]
```diff
@@ -96,7 +100,8 @@ export const generatePolygonInput = ({
   bBox,
   coordinates,
   crs,
-}: GenerateGeometryInput<PolygonInput['coordinates']> = {}): fc.Arbitrary<PolygonInput> => {
+  closeLinearRing,
+}: GenerateGeometryInput<PolygonInput['coordinates']> & PolygonInputOptions = {}): fc.Arbitrary<PolygonInput> => {
```
Comment (shimoncohen) [inline tests/geometries/helpers/geometries.ts:104]: Please open an issue about self intersecting geometries.

**Review by shimoncohen — APPROVED** [review]

File: tests/geometries/helpers/geometries.ts (hunk 13 above)
Comment (vitaligi) [reply tests/geometries/helpers/geometries.ts:104]: https://github.com/MapColonies/flat-earth/issues/24

**Review by shimoncohen — APPROVED** [review]

**Pull request merged**